### PR TITLE
implement csprng for numpy

### LINKED ISSUE #1164: Provide a CSPRNG utility for bindings languages

### PR DESCRIPTION
**Pull request by Shoeboxam**
Closes #1164

I'm sourcing noise from our Rust CSPRNG, for consistency.

### CODE REVIEW

**Comment by Shoeboxam** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/opendp/opendp/1165?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

Current dependencies on/for this PR:
* `main`
  * **PR #1162** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1162?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
    * **PR #1158** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1158?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
      * **PR #1155** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1155?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
      * **PR #1044** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1044?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
        * **PR #958** <a href="https://app.graphite.dev/github/pr/opendp/opendp/958?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
          * **PR #1147** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1147?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
            * **PR #1157** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1157?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
            * **PR #1141** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1141?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
              * **PR #1165** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1165?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈
                * **PR #1045** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1045?utm_source=stack-comment-icon" target="_blank">Graphite</a> 

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/opendp/opendp/1165?utm_source=stack-comment).

**Review by mccalluc — APPROVED** [review]
Since it seems like our use of numpy is pretty thin, just an interface really, would it be worth trying to make this work without that extra python dependency?

Is the test partly about seeing the performance when lots of randomness is needed? Or is 100,000 samples not really that many, but we're doing it to make the probability of a false failure very low? If it is kind of a stress test, I'm not sure it needs to be in every CI run.

**Comment by Shoeboxam** [comment]
@mccalluc I went ahead and added the dependency here, because I knew we'd be adding sklearn anyways in the next PR. I think our tests will need to be able to check PCA, and other models to come. 

The 100k samples runs almost instantaneously (.2 seconds), and it's only that high to make the type 1 error low.

**Comment by Shoeboxam** [comment]
Well, hey, scipy is a dependency of sklearn, so I might as well add it now.

File: .github/workflows/smoke-test.yml [hunk 1]
```diff
@@ -140,7 +140,7 @@ jobs:
           path: python/src/opendp/lib
       
       - name: Install package
-        run: python -m pip install -e .
+        run: python -m pip install -e .[scikit-learn]
```
Comment (mccalluc) [inline .github/workflows/smoke-test.yml:143]: Reminder to self:
- #1037

so latter on we can confirm in CI that it works both with and without the optional dependencies.

File: python/src/opendp/_data.py [hunk 2]
```diff
@@ -107,6 +108,38 @@ def ffislice_of_anyobjectptrs(
     return output
 
 
+@versioned
+def fill_bytes(
+    ptr,
+    len: int
+) -> bool:
+    r"""Internal function. Populate the buffer behind `ptr` with `len` random bytes
```
Comment (mccalluc) [inline python/src/opendp/_data.py:116]: Any argument for making it `_fill_bytes` if it's just internal?

Comment (mccalluc) [reply python/src/opendp/_data.py:116]: Never mind, this is generated.

**Review by mccalluc — APPROVED** [review]
Looks good to me.

File: python/src/opendp/_data.py (hunk 2 above)
Comment (Shoeboxam) [reply python/src/opendp/_data.py:116]: And it's in a private module!

**Pull request merged**